### ISSUE TITLE
Added extensions cause samtools error

### ISSUE THREAD

**Issue by Thomieh73** [sunbeam-labs/sunbeam#192]
Hi, today I tried to add the Kaiju extension to my sunbeam installation (which already had the extension sbx_report).

I got the same error as in the closed issue #181, where samtools is not finding libcrypto.so.1.0.0

Similar to the problem described i #181, I also had the channels in the wrong order. So I updated my ´.condarc`file. In the same way.

Then I removed the sunbeam conda environment as well as the sunbeam_stable directory, and started of with a new installation in the same way as I did originally.

my commands for the basic installation:
```
git clone -b stable https://github.com/eclarke/sunbeam sunbeam-stable
cd sunbeam-stable

./install.sh -c /work/projects/nn9305k/src/anaconda3/condabin/conda -v
source activate sunbeam
tests/run_tests.bash -e sunbeam  ##worked without problems
```

Next I added the two extensions: sbx_report and sbx_kaiju, while having the environment active, and from within the sunbeam_stable directory:
```
git clone https://github.com/sunbeam-labs/sbx_report extensions/sbx_report
conda install --file extensions/sbx_report/requirements.txt

git clone https://github.com/sunbeam-labs/sbx_kaiju extensions/sbx_kaiju
conda install --file extensions/sbx_kaiju/requirements.txt
```
and then I ran the test script`run_tests.bash` to test the installation as before. That failed at the step: rule align_to_host:

the error output:
```
samtools: error while loading shared libraries: libcrypto.so.1.0.0: cannot open shared object file: No such file or directory
Error in rule align_to_host:
    jobid: 104
    output: /tmp/tmp.QDFIJseR7i/sunbeam_output/qc/decontam/intermediates/human/dummyecoli.bam

RuleException:
CalledProcessError in line 38 of /work/projects/nn9305k/src/sunbeam-stable/rules/qc/decontaminate.rules:
Command ' set -euo pipefail;  
        bwa mem -M -t 1         /tmp/tmp.QDFIJseR7i/hosts/human.fasta         /tmp/tmp.QDFIJseR7i/sunbeam_output/qc/cleaned/dummyecoli_1.fastq.gz /tmp/tmp.QDFIJseR7i/sunbeam_output/qc/cleaned/dummyecoli_2.fastq.gz -o /tmp/tmp.QDFIJseR7i/sunbeam_output/qc/decontam/intermediates/human/dummyecoli.sam &&         samtools view -bSF4 /tmp/tmp.QDFIJseR7i/sunbeam_output/qc/decontam/intermediates/human/dummyecoli.sam > /tmp/tmp.QDFIJseR7i/sunbeam_output/qc/decontam/intermediates/human/dummyecoli.bam &&         rm /tmp/tmp.QDFIJseR7i/sunbeam_output/qc/decontam/intermediates/human/dummyecoli.sam ' returned non-zero exit status 127.
  File "/work/projects/nn9305k/src/sunbeam-stable/rules/qc/decontaminate.rules", line 38, in __rule_align_to_host
  File "/work/projects/nn9305k/src/anaconda3/envs/sunbeam/lib/python3.6/concurrent/futures/thread.py", line 56, in run
Removing output files of failed job align_to_host since they might be corrupted:
/tmp/tmp.QDFIJseR7i/sunbeam_output/qc/decontam/intermediates/human/dummyecoli.bam
Shutting down, this might take some time.
Exiting because a job execution failed. Look above for error message
Complete log: /work/projects/nn9305k/src/sunbeam-stable/.snakemake/log/2019-02-21T160723.826593.snakemake.log
x (log: /tmp/tmp.QDFIJseR7i/test_all.[out/err])
```
Any ideas on how to solve this?

btw: running the test without the `-e sunbeam` flag failed, altough it is described in your manual.

**Comment by eclarke** [comment]
It looks like this problem originates somewhere with Conda or Bioconda, and people from the original issue that @louiejtaylor found are still having problems: https://github.com/bioconda/bioconda-recipes/issues/12100

The solution appears to be explicitly downgrading openssl (the source of the libcrypto library):
```
conda install -c bioconda r-rcurl openssl=1.0
```
Would you try this and see if it works for you?

Thanks,
Erik

**Comment by Thomieh73** [comment]
Hi Eric,

I have tried the option you suggested. 
The first thing I did was update conda to the latest version (4.6.7) and then I installed the tools as before. And finally I forced the downgrading of openssl (v1.0). That resolved the samtools issue, so mapping worked again. The downside of that was that I had the next issue popping up. I could not run the final report option from thr sbx_report option. See log below.

```
[Sun Feb 24 21:58:35 2019] rule final_report:
[Sun Feb 24 21:58:35 2019]     input: /work/projects/nn9305k/air_sample/analysis/experiment_2/sunbeam_output/qc/reports/fastqc_quality.tsv, /work/projects/nn9305k/air_sample/analysis/experiment_2/sunbeam_output/qc/reports/preprocess_summary.tsv, /work/projects/nn9305k/air_sample/analysis/experiment_2/sunbeam_output/classify/kraken/all_samples.tsv
[Sun Feb 24 21:58:35 2019]     output: /work/projects/nn9305k/air_sample/analysis/experiment_2/sunbeam_output/reports/final_report.html
[Sun Feb 24 21:58:35 2019]     jobid: 0

Quitting from lines 51-70 (.snakemake.lc7quues.final_report.Rmd) 
Error: package or namespace load failed for 'tidyverse' in dyn.load(file, DLLpath = DLLpath, ...):
 unable to load shared object '/work/projects/nn9305k/src/anaconda3/envs/sunbeam/lib/R/library/haven/libs/haven.so':
  /work/projects/nn9305k/src/anaconda3/envs/sunbeam/lib/R/library/haven/libs/haven.so: undefined symbol: libiconv
Execution halted
[Sun Feb 24 21:58:39 2019] Error in rule final_report:
[Sun Feb 24 21:58:39 2019]     jobid: 0
[Sun Feb 24 21:58:39 2019]     output: /work/projects/nn9305k/air_sample/analysis/experiment_2/sunbeam_output/reports/final_report.html

[Sun Feb 24 21:58:39 2019] RuleException:
[Sun Feb 24 21:58:39 2019] CalledProcessError in line 9 of /work/projects/nn9305k/src/sunbeam-stable/extensions/sbx_report/sbx_report.rules:
[Sun Feb 24 21:58:39 2019] Command ' set -euo pipefail;  Rscript -e 'rmarkdown::render("/work/projects/nn9305k/src/sunbeam-stable/extensions/sbx_report/.snakemake.lc7quues.final_report.Rmd", output_file="/work/projects/nn9305k/air_sample/analysis/experiment_2/sunbeam_output/reports/final_report.html", quiet=TRUE, params = list(rmd="/work/projects/nn9305k/src/sunbeam-stable/extensions/sbx_report/.snakemake.lc7quues.final_report.Rmd"))' ' returned non-zero exit status 1.
[Sun Feb 24 21:58:39 2019]   File "/work/projects/nn9305k/src/sunbeam-stable/extensions/sbx_report/sbx_report.rules", line 9, in __rule_final_report
[Sun Feb 24 21:58:39 2019]   File "/work/projects/nn9305k/src/anaconda3/envs/sunbeam/lib/python3.6/concurrent/futures/thread.py", line 56, in run
[Sun Feb 24 21:58:39 2019] Shutting down, this might take some time.
[Sun Feb 24 21:58:39 2019] Exiting because a job execution failed. Look above for error message
[Sun Feb 24 21:58:39 2019] Complete log: /work/projects/nn9305k/air_sample/analysis/.snakemake/log/2019-02-24T215750.136081.snakemake.log
```
So I decided to switch around the installation of the extensions. First install kaiju and then the sbx_reports, just to see if that was doing anything different. That was not the case.

I have attached the reports from each installation:
The kaiju installation

```
The following NEW packages will be INSTALLED:

  gnu-wget           bioconda/linux-64::gnu-wget-1.18-3
  kaiju              bioconda/linux-64::kaiju-1.6.3-hdbcaa40_1

The following packages will be UPDATED:

  openssl                                 1.0.2q-h14c3975_0 --> 1.1.1a-h14c3975_1000
  python             conda-forge::python-3.6.7-hd21baee_10~ --> pkgs/main::python-3.6.8-h0371630_0

The following packages will be SUPERSEDED by a higher-priority channel:

  cryptography       conda-forge::cryptography-2.5-py36hb7~ --> pkgs/main::cryptography-2.5-py36h1ba5d50_0
  curl                  conda-forge::curl-7.64.0-h646f8bb_0 --> pkgs/main::curl-7.62.0-hbc83047_0
  libcurl            conda-forge::libcurl-7.64.0-h01ee5af_0 --> pkgs/main::libcurl-7.62.0-h20c2e04_0

The following packages will be DOWNGRADED:

  krb5                                 1.16.3-hc83ff2d_1000 --> 1.14.6-0
  libssh2                               1.8.0-h1ad7b7a_1003 --> 1.8.0-1
  samtools                                   1.9-h57cc563_6 --> 1.9-h91753b0_5
  wget                                    1.19.5-h1ad7b7a_0 --> 1.19.5-h20c2e04_0
```
The sbx_report installation

```
The following packages will be UPDATED:

  krb5                           conda-forge::krb5-1.14.6-0 --> pkgs/main::krb5-1.16.1-h173b8e3_7
  pcre                 conda-forge::pcre-8.41-hf484d3e_1003 --> pkgs/main::pcre-8.42-h439df22_0

The following packages will be SUPERSEDED by a higher-priority channel:

  fontconfig         conda-forge::fontconfig-2.13.1-h2176d~ --> pkgs/main::fontconfig-2.13.0-h9420a91_0

The following packages will be DOWNGRADED:

  libuuid                              2.32.1-h14c3975_1000 --> 1.0.3-1
```

and finally the downgrading of openssl:

```
The following NEW packages will be INSTALLED:

  r-bitops           conda-forge/linux-64::r-bitops-1.0_6-r351h96ca727_1002
  r-rcurl            conda-forge/linux-64::r-rcurl-1.95_4.11-r351h96ca727_1003

The following packages will be UPDATED:

  cryptography       pkgs/main::cryptography-2.5-py36h1ba5~ --> conda-forge::cryptography-2.5-py36hb7f436b_1
  curl                    pkgs/main::curl-7.62.0-hbc83047_0 --> conda-forge::curl-7.64.0-h646f8bb_0
  krb5                    pkgs/main::krb5-1.16.1-h173b8e3_7 --> conda-forge::krb5-1.16.3-hc83ff2d_1000
  libcurl              pkgs/main::libcurl-7.62.0-h20c2e04_0 --> conda-forge::libcurl-7.64.0-h01ee5af_0
  r-openssl          pkgs/r::r-openssl-1.0.2-r351h96ca727_1 --> conda-forge::r-openssl-1.1-r351hff1dc39_1001

The following packages will be SUPERSEDED by a higher-priority channel:

  python                 pkgs/main::python-3.6.8-h0371630_0 --> conda-forge::python-3.6.7-hd21baee_1002

The following packages will be DOWNGRADED:

  openssl                              1.1.1a-h14c3975_1000 --> 1.0.2q-h14c3975_0
  wget                                    1.19.5-h20c2e04_0 --> 1.19.5-h1ad7b7a_0
```
If I see it right, the downgrading gives a problem with the sbx_report script, where the tidyverse can not find a library. 

My idea now is that if I want to use sbx_kaiju, I can not use sbx_report and vice versa. Am I correct there.

**Comment by eclarke** [comment]
Yes, unfortunately it seems like you're correct. The way to get around this is to use per-rule conda environments, which we haven't been doing for these extensions. Here is more information on that: https://snakemake.readthedocs.io/en/stable/snakefiles/deployment.html#integrated-package-management. Ideally this would mean that each extension can install its requirements in a new conda environment that activates "on top of" the sunbeam environment, and hopefully would not conflict. This conda environment would only activate when rules from the extension are called.

I will file issues with the two extensions you're referencing to fix that and ping @ressy and @louiejtaylor so we can start updating.

**Comment by ressy** [comment]
Thanks Erik.  We use that approach for https://github.com/sunbeam-labs/sbx_metaphlan and it seems to work smoothly so I expect it'll be the same for these other two.  I'll take a look.

**Comment by ressy** [comment]
This looks straightforward for sbx_report, but sbx_kaiju is a little trickier since you need to be able to use kaiju's tools interactively for the initial database setup.  It works if you already have those files ready but I'm still thinking of what the instructions should recommend when starting from scratch.

But Thomas, for your immediate problem here, you could give it a try with these modified extensions if you already have your kaiju database ready.  I have a "use-conda" branch on each of these with the changes:

https://github.com/ressy/sbx_kaiju
https://github.com/ressy/sbx_report

Basically just skip the `conda install --file` step (no more requirements.txt) and add `--use-conda` when calling snakemake.  Then it should automatically maintain isolated environments based on the .yml file in each extension directory.

Jesse

**Comment by ArwaAbbas** [comment]
> 
> 
> It looks like this problem originates somewhere with Conda or Bioconda, and people from the original issue that @louiejtaylor found are still having problems: [bioconda/bioconda-recipes#12100](https://github.com/bioconda/bioconda-recipes/issues/12100)
> 
> The solution appears to be explicitly downgrading openssl (the source of the libcrypto library):
> 
> ```
> conda install -c bioconda r-rcurl openssl=1.0
> ```
> 
> Would you try this and see if it works for you?
> 
> Thanks,
> Erik

I've also had an issue with re-installing sunbeam (with a new environment) and then not having the "all_mapping" work due to the following:

`samtools: error while loading shared libraries: libcrypto.so.1.0.0: cannot open shared object file: No such file or directory`

However, installing opennssl in the sunbeam environment (as Erik describes above) seems to fix that.

**Comment by louiejtaylor** [comment]
@eclarke/@ressy Chris Gu also had the libcrypto.so issue with a fresh install today. Should we pin `openssl=1.0` in the standard Sunbeam env?

**Comment by eclarke** [comment]
I guess so. Let’s at least try it with the integration tests.

**Comment by Thomieh73** [comment]
Hey @ressy I will try out your suggestion next week. I keep you updated
Thomas

**Comment by ressy** [comment]
Thanks @Thomieh73, sounds good.

@eclarke / @louiejtaylor / @ArwaAbbas I'm seeing if I can reproduce the openssl versioning issue myself in a new isolated setup (with a separate HOME and such).  Thinking out loud a bit: Ideally shouldn't our existing integration tests be catching it already, like during test_all?  I wonder if the ~/miniconda3 cache in CircleCI is preventing the problem from surfacing since it doesn't start with the latest available set of packages, and maybe we should have a periodic cron job scheduled using a cache-less version.  But if we do pin openssl to 1.0 that'd be a moot point for this particular case.

**Comment by eclarke** [comment]
In general, I think a period cache invalidation would be a good idea to prevent bit rot, unless we just pin _all_ the requirements to their specific versions, but that make things difficult with extensions.

By the way, this issue seems related to the conda version the user is running. The bioconda team is  working on a resolution to this in bioconda and conda-forge: https://github.com/bioconda/bioconda-recipes/issues/13958

**Comment by ressy** [comment]
Aha thanks Erik, that and the other issue you linked help me see why I haven't been able to reproduce it now from scratch.  Since they're actively fixing it to make things work with openssl 1.1, maybe we shouldn't pin it at 1.0 after all?

In general, I'd still lean toward pinning exact versions only on a case-by-case basis rather than across-the-board.  Maybe we could just set the cache key to include the current year+month or similar, so it'll automatically reset every so often.  Again, just thinking out loud, though.

**Comment by ressy** [comment]
I just thrashed this out in a VM to try to pin down the samtools/openssl issue once and for all.  I'll put my description here since this is our remaining open issue that is at least partially related and has a lot of our existing discussion on it.  **tl;dr: Bioconda's samtools builds from samtools=1.9=h8571acd_10 onward seem OK to me.  Earlier ones are screwy.**

The bugs (I think there are two separate problem depending on the exact build) are in the bioconda package builds for samtools in a bunch of earlier iterations over the past few months, including most of the ones for 1.9.  For some builds the samtools binaries link against libcrypto 1.0 but the conda packages for those builds only pull in 1.1 via package dependencies.  I think these are just too loose in their listed dependencies so you could end up satisfying the libcrypto dependency with openssl=1.1 (and breaking samtools) even though it really should have explicitly required 1.0.  Other builds correct the dependency and *do* correctly insist on 1.0, but then of course we have trouble with other packages that want 1.1.  And the dependencies for _9 seem completely broken for some reason.  None of this behavior relates to updates on conda-forge particularly; it comes up with just bioconda's samtools and the default channels for dependencies.

I think these are the key points:

 * If you start from scratch with new Miniconda and Sunbeam and everything, it all works fine, as we've been seeing.
 * If you have an earlier samtools build installed *and* you don't also just happen to have the right conda-provided openssl installed *and* your OS doesn't provide a `libcrypto.so` for the right version under a standard library path, then you'll encounter this problem.  That last aspect was why I was having so much trouble reproducing this.
 * Some attempted fixes I see recommended elsewhere like `conda install samtools` don't work because the package manager thinks the latest (1.9) is already installed and doesn't care about the different exact builds unless you force it.

No wonder there's so much confusion floating around online about it.  But I think the takeaway may just be to install bioconda's latest samtools build and let it pull in its openssl 1.1 dependency and call it a day.

-------------------

**Back to the original problem here:**  I'm thinking @Thomieh73's problem will have gone away since that was dealing with the earlier broken builds.  If anything is still needed `conda install -c bioconda samtools=1.9=h8571acd_11` should do the trick, if I have this right.

**Comment by ressy** [comment]
I just tested the `--force-reinstall` (available in recent conda versions) mentioned in the bioconda issue and it *does* recognize the latest build as the newest available package.  So `conda install --force-reinstall -c bioconda samtools` should default to the very latest and fix it.

**Comment by louiejtaylor** [comment]
Huge thank you to @ressy for tracking down the source of this problem and for coming up with a simple solution!

**Comment by ressy** [comment]
I'm going to resolve this one since we know what to do now and new installs should be OK.  But @Thomieh73, let us know if you still have trouble.
